indexserver: port merging policy to Go

stefanhengl opened the pull request:
This ports the merging policy in /cmd/zoekt-merge-index from Python to
Go. I added the ported policy to zoekt-sourcegraph-indexserver because
this is were merging will finally run in a background process.

To make testing easier, I added a command to indexserver that lets us
trigger a merge job manually, just like we do with the python script.
The way I treat parameters is a bit hacky but this way don't have to add
a ton of parameters to the flagset and remove them later on.

After testing the merge policy manually, the next step is to run merging
every x hours on a subset of our instances in production and monitor the
behavior.

**Summary of the merging policy:**
All shards are eligable for merging except:
- repos with recent commits (last 7 days)
- popular repos (ranking greater than 100)
- large repos (repo that are distributed accross many shards)

Explicity, compound shards are eligable for merging, too. Compound shards
shrink over time, because we vacuum the tombstones, and the smaller
they get the higher the probabiliy that are subject to a merge.

The merge policy tries to build compund shard that match the desired 
target size.

stefanhengl committed:
indexserver: port merging policy to Go

This ports the merging policy in /cmd/zoekt-merge-index from Python to
Go. I added the ported policy to zoekt-sourcegraph-indexserver because
that is were merging will finally run in background process.

To make testing easier, I added a command to indexserver that let's us
trigger a merge job manually.

After testing the merge policy manually, the next step is to run merging
every x hours on a subset of our instances in production and monitor the
behavior.

stefanhengl committed:
simplify cleanup

stefanhengl committed:
simplify parsing parameters

stefanhengl committed:
more doctstrings

stefanhengl committed:
import strings again

keegancsmith committed:
include debugMerge in debugCmd list

keegancsmith commented:
reviewing now, just merged in master and made a small change based on something else I changed in indexserver.

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/merge.go:
```diff
@@ -0,0 +1,237 @@
+	}
+	debug.Printf("found %d shards\n", len(shards))
+
+	backupDir := filepath.Join(dir, ".scratch/bak")
```
do we need to make dir configurable? It is an argument that is passed in to indexserver.

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/merge.go:
```diff
@@ -0,0 +1,237 @@
+	shards = shards[cur:]
+
+	sort.Slice(shards, func(i, j int) bool {
+		return shards[i].Rank < shards[j].Rank
```
Should we maybe consider higher priority shards first, so they end up at the front of the compound shard? I quite like that we merge unimportant shards first. So maybe sort the compound afterwards so it is in decreasing rank order.

keegancsmith commented on cmd/zoekt-sourcegraph-indexserver/merge.go:
```diff
@@ -0,0 +1,237 @@
+			continue
+		}
+	}
+	if cur == len(shards) {
```
nit: maybe check below that `shards` is empty. Easier to validate correct since you don't need to understand chat cur == len(shards) means.

stefanhengl committed:
rewrite policy

stefanhengl committed:
simplify generateCompounds

stefanhengl committed:
minor fixes

stefanhengl commented:
I updated the implementation. Most importantly, I moved the burden of deciding which shards get merged and which don't from `generateCompounds` to `loadCandidates` (previously `loadShards`). Now there is also a dedicated function `isExcluded` that makes it easier to update the merging policy. 

For the review, I think it is easiest to start with `doMerge`, everything else is called from there.

stefanhengl committed:
fix priority: int -> float64

keegancsmith reviewed (approved):
LGTM. 

My main concern is around ranking. I see you exclude repos bigger than 100 stars, so this should work quite well. I think another PR which makes zoekt-merge merge in order of priority will be quite useful to get some better behaviour.

Is there a way we can group things by rank? I know you had something like that.

stefanhengl commented:
>My main concern is around ranking. I see you exclude repos bigger than 100 stars, so this should work quite well. I think another PR which makes zoekt-merge merge in order of priority will be quite useful to get some better behaviour.

I agree. I felt like I had to make a trade-off. The approach in this PR has the nice property that it will try to refill the compound shards with new repos before creating new compound shards. Since the size of a compound shard is a good predictor of the gains we get, I think this is reasonable.
 
> Is there a way we can group things by rank? I know you had something like that.

This sounds like an optimization problem where we want to minimize `alpha * var(rank)+ beta * abs(target size - actual size)`.

stefanhengl committed:
move up hasMultipleShards

stefanhengl committed:
prefix logs with merging:

stefanhengl committed:
parse cmd outpu and log merge actions to tsv

stefanhengl committed:
fix logging, better docstrings
